Infinite loop with by name value classes

```scala
class Foo(sc: => Int) extends AnyVal
```

emits the error 
```scala
2 |  class Foo(sc: => Int) extends AnyVal
  |        ^
  |        value sc in class Foo does not take parameters
```

but then stays in an infinite loop.

In scalac it fails with 
```
Foo.scala:2: error: value class parameter must be a val and not be private[this]
```

In both scalac and dotty 
```scala
class Foo(val sc: => Int) extends AnyVal
```
fails correctly with `val parameters may not be call-by-name`

--- a/compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java
+++ b/compiler/src/dotty/tools/dotc/reporting/diagnostic/ErrorMessageID.java
@@ -132,7 +132,8 @@ public enum ErrorMessageID {
     ImportRenamedTwiceID,
     TypeTestAlwaysSucceedsID,
     TermMemberNeedsNeedsResultTypeForImplicitSearchID,
-    CaseClassCannotExtendEnumID
+    CaseClassCannotExtendEnumID,
+    ValueClassParameterMayNotBeCallByNameID
     ;
 
     public int errorNumber() {

--- a/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala
@@ -1666,6 +1666,13 @@ object messages {
     val explanation = ""
   }
 
+  case class ValueClassParameterMayNotBeCallByName(valueClass: Symbol, param: Symbol)(implicit ctx: Context)
+    extends Message(ValueClassParameterMayNotBeCallByNameID) {
+    val msg = s"value class parameter `${param.name}` may not be call-by-name"
+    val kind = "Syntax"
+    val explanation = ""
+  }
+
   case class OnlyCaseClassOrCaseObjectAllowed()(implicit ctx: Context)
     extends Message(OnlyCaseClassOrCaseObjectAllowedID) {
     val msg = "only `case class` or `case object` allowed"

--- a/compiler/src/dotty/tools/dotc/typer/Checking.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Checking.scala
@@ -537,6 +537,8 @@ object Checking {
           case param :: params =>
             if (param.is(Mutable))
               ctx.error(ValueClassParameterMayNotBeAVar(clazz, param), param.pos)
+            if (param.info.isInstanceOf[ExprType])
+              ctx.error(ValueClassParameterMayNotBeCallByName(clazz, param), param.pos)
             if (param.is(Erased))
               ctx.error("value class first parameter cannot be `erased`", param.pos)
             else {

--- a/compiler/test/dotty/tools/dotc/reporting/ErrorMessagesTests.scala
+++ b/compiler/test/dotty/tools/dotc/reporting/ErrorMessagesTests.scala
@@ -935,6 +935,18 @@ class ErrorMessagesTests extends ErrorMessagesTest {
       assertEquals("class MyValue", valueClass.show)
     }
 
+  @Test def valueClassParameterMayNotBeCallByName =
+    checkMessagesAfter(RefChecks.name) {
+      """class MyValue(a: => Int) extends AnyVal"""
+    }
+    .expect { (ictx, messages) =>
+      implicit val ctx: Context = ictx
+      assertMessageCount(1, messages)
+      val ValueClassParameterMayNotBeCallByName(valueClass, param) :: Nil = messages
+      assertEquals("class MyValue", valueClass.show)
+      assertEquals("value a", param.show)
+    }
+
   @Test def onlyCaseClassOrCaseObjectAllowed =
     checkMessagesAfter(FrontEnd.name) {
       """case Foobar"""

